### PR TITLE
[feature](table-valued-function) Support S3 tvf

### LINKED ISSUE #10217: [Feature] support table valued function.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

A table-valued function (TVF) is a function that returns a relation or a set of rows. There are two types of TVFs :
- a TVF that can be specified in a FROM clause, e.g. range;
- a TVF that can be specified in SELECT/LATERAL VIEW clauses, e.g. explode.

Just like the definition of spark and clickhouse：
- spark: https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-tvf.html.
- clickhouse: https://clickhouse.com/docs/en/sql-reference/table-functions.  Currently Clickhouse only support from clause table function

The table valued function can be used to test our vector engines performance.

### Use case

The table valued function works like this：
MySQL [(none)]> select sum(number) from (select number from numbers("1000") where number < 10) as c;
+---------------+
| sum(number) |
+---------------+
| 45 |
+---------------+

That is: numbers("1000") returns a temporary table which only has one column named 'number' and it's row-values are from 0 to 1000.

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by BePPPower**
# Proposed changes

Issue Number: close #10217
pr Number: close #10214

This pr does three things：

1. Modified the framework of table-valued-function(tvf).
2. be support `fetch_table_schema` rpc.
3. Implemented `S3(path, AK, SK, format)` table-valued-function.

## Problem summary

1. Modified the framework of table-valued-function(tvf).
Each implementation of tvf requires a new `scan node` in the old framework of pr #10214 . Actually, different `scan nodes` should correspond to different data sources. I modified the tvf framework so that it can generate different `scan node` according to data source of this tvf.
For example, the data of `numbers(n,m)` tvf is derived from memory, so it generate `DataGenScanNode`. And the data of `s3(...)` tvf is derived from S3 Storage, so it generate `ExternalFileScanNode`.

2. be support `fetch_table_schema` rpc.
There is no table schema in `s3()` advance, so we need `fetch_table_schema` rpc to parse table schema from file in S3 Storage.

3. `S3(path, AK, SK, format)`
The initial implement of `s3(..)` tvf, use like:
```
MySQL [(none)]> select * from s3("http://127.0.0.1:9312/test2/student1.csv", "minioadmin", "minioadmin", "csv");
+------+------+------+
| c1   | c2   | c3   |
+------+------+------+
| 1    | ftw  | 12   |
| 2    | zs   | 18   |
| 3    | ww   | 20   |
+------+------+------+
3 rows in set (0.04 sec)

MySQL [(none)]> select * from s3("http://127.0.0.1:9312/test2/student_with_names.csv", "minioadmin", "minioadmin", "csv_with_names");
+------+------+------+
| id   | name | age  |
+------+------+------+
| 1    | ftw  | 12   |
| 2    | zs   | 18   |
| 3    | ww   | 20   |
+------+------+------+
3 rows in set (0.04 sec)

```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
5. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
6. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
7. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
8. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/service/internal_service.cpp [hunk 1]
```diff
@@ -407,6 +410,78 @@ void PInternalServiceImpl::fetch_data(google::protobuf::RpcController* cntl_base
...
+    }
+    result->set_column_nums(col_names.size());
+    for (size_t idx = 0; idx < col_names.size(); ++idx) {
+        result->add_column_names(col_names[idx]);
+    }
+    for (size_t idx = 0; idx < col_types.size(); ++idx) {
+        PTypeDesc* type_desc = result->add_column_types();
+        col_types[idx].to_protobuf(type_desc);
+    }
+    LOG(INFO) << "complete parse, status: " << st;
```
Comment (morningman) [inline be/src/service/internal_service.cpp:481]: Remove this log, meaningless

File: be/src/service/internal_service.cpp [hunk 2]
```diff
@@ -407,6 +410,78 @@ void PInternalServiceImpl::fetch_data(google::protobuf::RpcController* cntl_base
...
+    std::vector<SlotDescriptor*> file_slots;
+    std::unique_ptr<vectorized::GenericReader> reader(nullptr);
+    std::unique_ptr<RuntimeProfile> profile(new RuntimeProfile("FetchTableSchema"));
+    switch (params.format_type) {
+    case TFileFormatType::FORMAT_CSV_PLAIN:
+    case TFileFormatType::FORMAT_CSV_GZ:
+    case TFileFormatType::FORMAT_CSV_BZ2:
+    case TFileFormatType::FORMAT_CSV_LZ4FRAME:
+    case TFileFormatType::FORMAT_CSV_LZOP:
+    case TFileFormatType::FORMAT_CSV_DEFLATE: {
```
Comment (morningman) [inline be/src/service/internal_service.cpp:454]: Missing Parquet and ORC?

File: fe/fe-core/src/main/java/org/apache/doris/planner/DataGenScanNode.java [hunk 3]
```diff
@@ -40,15 +40,15 @@
 /**
  * This scan node is used for table valued function.
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DataGenScanNode.java:41]: Change the comment

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java [hunk 4]
```diff
@@ -202,6 +187,39 @@ public void init(Analyzer analyzer) throws UserException {
...
+                break;
+            default:
+                throw new UserException("Unknown table type: " + hmsTable.getDlaType());
+        }
+        this.scanProviders.add(scanProvider);
+    }
+
+    private void initFunctionGenTable(FunctionGenTable table, ExternalFileTableValuedFunction tvf) {
+        Preconditions.checkNotNull(table);
+        FileScanProviderIf scanProvider;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java:218]:
```suggestion
        FileScanProviderIf scanProvider = new TVFScanProvider(table, desc, tvf);;
```

File: be/src/service/internal_service.cpp (hunk 2 above)
Comment (BePPPower) [reply be/src/service/internal_service.cpp:454]: Parquet and ORC will be added after implementing `get_parsered_schema` method of  ParquetReader and OrcReader.

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by morningman** [comment]
Need to add some regression test in next PR

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.44 seconds
 load time: 432 seconds
 storage size: 17154655363 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221104103500_clickbench_pr_40058.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**